spec/std/digest/algorithm.cr

```diff
@@ -0,0 +1,47 @@
+macro acts_as_digest_algorithm(expr)
+  it "#hexdigest can update within a loop from explicit expr (#9483)" do
+    %i = 0
+    {{expr}}.hexdigest do |%digest|
+      while %i < 3
+        %digest.update("")
+        %i += 1
+      end
+    end
+  end
+
+  it "#hexdigest can update within a loop by indirect expr (#9483)" do
+    %algorithm = {} of String => Digest::Algorithm.class
+    %algorithm["me"] = {{expr}}
+    %i = 0
+    %algorithm["me"].hexdigest do |%digest|
+      while %i < 3
+        %digest.update("")
+        %i += 1
+      end
+    end
+  end
+
+  it "context are independent" do
+    %algorithm = {{expr}}
+    %res = %algorithm.hexdigest do |%digest|
+      %digest.update("a")
+      %digest.update("b")
+    end
+
+    %inner_res = nil
+
+    %outer_res = %algorithm.hexdigest do |%outer|
+      %outer.update("a")
+
+      %inner_res = %algorithm.hexdigest do |%inner|
+        %inner.update("a")
+        %inner.update("b")
+      end
+
+      %outer.update("b")
+    end
+
+    %outer_res.should eq(%res)
+    %inner_res.should eq(%res)
+  end
+end
```

Comment:
There's no need for fresh variables in here, since they're scoped within `it` blocks anyway.

Reply:
It depends the value of expr. For example at some point I used the macro as 
`acts_as_digest_algorithm algorithm` in the OpenSSL::Digest specs which clash with  `%algorithm` unless `%` is used.

Reply:
Actually, the macro may not even be needed – just a method. `forall T.class` can be required though.

I am not against macros, but it would be definitely easier to debug to use a `def`. Of course, `line: line` has to be added to `it`s.

Reply:
Yes, having this as a macro seems completely unnecessary for how this is used.

Reply:
I think it emulates better what the program typed by the user would be. Because I care about the expression and not the value. Since the end result is that this is used with types, then we can change it to a def, ok.